### PR TITLE
Feat/RwaInputConduit3 (MIP65)

### PR DESCRIPTION
**Pull request by nduchak**
[sc-957]

Create RwaInputConduit3.This contract differs from the original [RwaInputConduit](https://github.com/makerdao/MIP21-RWA-Example/blob/fce06885ff89d10bf630710d4f6089c5bba94b4d/src/RwaConduit.sol#L20-L39):
 - The caller of `push()` is not required to hold MakerDAO governance tokens.
 - The `push()` method is permissioned.
 - `push()` permissions are managed by `mate()`/`hate()` methods.
 - Require PSM address in the constructor
 - The `push()` method swaps GEM to DAI using PSM
- The `quit` method allows moving outstanding GEM balance to `quitAddress`. It can be called only by the admin.
- The `file` method allows updating `quitAddress`. It can be called only by the admin.

### CODE REVIEW

File: .gitignore
Comment (hbarcelos) [inline .gitignore:8]: I don't think want this here. This file is going to be required if we ever use any dependency which does not follow the dapp.tools/foundry directory structure.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: Since these 2 are not present in other conduits, they could be made `immutable` so they don't take up storage space.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: We can get all dependencies from the PSM contract instead of requiring them to be passed as parameters:

```suggestion
    constructor(address _psm, address _to) public {
        psm = PsmAbstract(_psm);
        dai = PsmAbstract(_psm).dai();
        gem = GemJoinAbstract(PsmAbstract(_psm).gemJoin()).gem();
        to = _to;

        wards[msg.sender] = 1;
        emit Rely(msg.sender);
    }
```

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: I think we should probably not add this method.
We we want to control the amount of Dai being disbursed, we should probably do it in the `urn.draw()` step, not in the conduit.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: Instead of approving on-demand, we should probably have an unlimited approval in the constructor to improve gas usage.

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.t.sol:275]: I believe these 2 lines are not executed because `inputConduit.push()` reverts.

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.t.sol]: Maybe rename this to `joinA`.

**Review by dbkbali — APPROVED** [review]
Looks good @nduchak subject to changes proposed by @hbarcelos - external to this and maybe you discussed this yesterday - @lollike - what about edge cases if they start charging fees or implementing negative rates - see this https://github.com/makerdao/dss-psm/pull/6 maybe something that doesn't need to be addressed in the conduit- but in the architecture

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.t.sol]: Are you sure anything after the first failure is even executing?

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: ⚠️⚠️⚠️ Careful here!!! ⚠️⚠️⚠️

`sellGem()` 2nd parameter is `gemAmt`, not `wad`. The precision is the one of the gem (i.e.: 6 decimals for USDC).

**Review by hbarcelos — CHANGES_REQUESTED** [review]

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: We probably don't need this guard here.
If someone calls `psm.sellGem` with `0`, it will simply be a no-op.

File: .gitignore
Comment (nduchak) [reply .gitignore:8]: VsCode requires `remappings.txt` to be able to map imports in Solidity files, that's the main reason why I put it there.
In case if we need custom remappings we can remove it from gitignore

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: As i remember `immutable` takes storage slot as well.

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Good catch

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Done

File: src/conduits/RwaInputConduit3.t.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol:275]: Actually, it is executed, `vm.expectRevert` catch revert from the function call

File: src/conduits/RwaInputConduit3.t.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol]: Done

File: src/conduits/RwaInputConduit3.t.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol]: Yeap

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Ok, just renamed argument to `gemAmount`

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: I put it there just to save some gas, as it will fail on this line: https://github.com/makerdao/dss-psm/blob/master/src/psm.sol#L113

Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: > Looks good @nduchak subject to changes proposed by @hbarcelos - external to this and maybe you discussed this yesterday - @lollike - what about edge cases if they start charging fees or implementing negative rates - see this [makerdao/dss-psm#6](https://github.com/makerdao/dss-psm/pull/6) maybe something that doesn't need to be addressed in the conduit- but in the architecture

Another edge case which we can have is DC of PSM, if we try to swap amount above DC then it will fail and our USDC will be stacked in conduit. Does it make sense to have a method to move USDC out from conduit???

File: src/conduits/RwaInputConduit3.t.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol:275]: That's why we don't have Fail in test name

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol]: > I put it there just to save some gas, as it will fail on this line: https://github.com/makerdao/dss-psm/blob/master/src/psm.sol#L113

Why would it fail in that line?

It's actually:
```
gemJoin.join(address(this), gemAmt, msg.sender);
```

For USDC we use [join-5-auth](https://github.com/makerdao/dss-psm/blob/7d717d88d6b6dfcb346119369647969df6cadad3/src/join-5-auth.sol). There is this require statement on L84:
```
require(int256(wad) >= 0, "AuthGemJoin5/overflow");
```

**Comment by hbarcelos** [comment]
> Another edge case which we can have is DC of PSM, if we try to swap amount above DC then it will fail and our USDC will be stacked in conduit. Does it make sense to have a method to move USDC out from conduit???

Yeah, that can become problematic. We should discuss this next week.

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.t.sol]: Yep, you are correct:
https://book.getfoundry.sh/cheatcodes/expect-revert

> You can also have multiple expectRevert() checks in a single test.
> 
> ```solidity
> function testMultipleExpectReverts() public {
>     vm.expectRevert(abi.encodePacked("INVALID_AMOUNT"));
>     vault.send(user, 0);
> 
>     vm.expectRevert(abi.encodePacked("INVALID_ADDRESS"));
>     vault.send(address(0), 200);
> }
> ```

TIL

**Comment by nduchak** [comment]
> > Another edge case which we can have is DC of PSM, if we try to swap amount above DC then it will fail and our USDC will be stacked in conduit. Does it make sense to have a method to move USDC out from conduit???
> 
> Yeah, that can become problematic. We should discuss this next week.

We already have some discussion with @lollike , he proposes to add `quit` function which simply moves outstanding GEM to`quitAddress` which can be changed by Governance

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: Maybe now inline this method in `push`, since it's called only once.

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.t.sol:275]: Sorry, you are right.

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Done

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.t.sol]: `WAD` literally means `10**18`. We should probably not be using this term here.
```suggestion
    uint256 constant USDX_BASE_UNIT = 10 ** 6;
```

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.t.sol]: This constant is not used anywhere else.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: @Probably this way of writing it is more idiomatic and probably more gas efficient (2**256 does not fit into a single word, so unless the compiler is optimizing this away, it will take 2 words during execution).

```suggestion
        _gem.approve(address(PsmAbstract(_psm).gemJoin()), type(uint256).max);
```

File: src/conduits/RwaInputConduit3.t.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol]: `wad`: some quantity of tokens, usually as a fixed point integer with 10^18 decimal places.

Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol]: This just an amount of some token, but yes this looks more clear

Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.t.sol]: Hmmm... yeah, that's the definition on the Glossary.
Maybe I'm being too nitpicky since this is just a test file, but when it comes to actual code, anything different that 10**18 is called something else (see the different join adapters for instance).

IDK, I'll leave it up to you.

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Yes, I mean this one if you go deeper.
We save some gas skipping these operations:
PSM:
```
uint256 gemAmt18 = mul(gemAmt, to18ConversionFactor);
uint256 fee = mul(gemAmt18, tin) / WAD;
uint256 daiAmt = sub(gemAmt18, fee);
```

GemJoin
```
 require(live == 1, "AuthGemJoin5/not-live");
 uint256 wad = mul(amt, 10 ** (18 - dec));
```

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: DOne

File: src/conduits/RwaInputConduit3.t.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol]: Done

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol]: I see...

Another point is that having this check actually deviates from the behavior of the other conduits, which do not fail when calling `push` with balance `0`:
https://github.com/clio-finance/mip21-toolkit/blob/3fdb932f6daf966223643add1e191397461f7c9a/src/conduits/RwaInputConduit2.sol#L130-L137

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol]: From the [docs](https://docs.soliditylang.org/en/v0.8.16/contracts.html?highlight=immutable#immutable):

> Constant and Immutable State Variables[](https://docs.soliditylang.org/en/v0.8.16/contracts.html?highlight=immutable#constant-and-immutable-state-variables)
>
> State variables can be declared as constant or immutable. In both cases, the variables cannot be modified after the contract has been constructed. For constant variables, the value has to be fixed at compile-time, while for immutable, it can still be assigned at construction time.
>
> It is also possible to define constant variables at the file level.
>
> **The compiler does not reserve a storage slot for these variables, and every occurrence is replaced by the respective value.**

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol:110]: Missing `_quitAddress` here.

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: I remember at one of the debugging spell session we find out out that `immutable` actually reserve slot

Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Not sure maybe I am wrong

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Hmm... Is it something which we should follow? If yes it's just optimization and we can remove it.

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol:110]: Done

Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol:110]: Did you push? I can't see it.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol]: Probably this is something we want to keep consistent across implementations. That `push()` method is the primary function of this contract, it shouldn't behave on unexpected ways.

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.t.sol]: Typo:

```suggestion
    event Quit(address indexed quitAddress, uint256 wad);
```

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: Typo:

```suggestion
 *  - The `file` method allows updating `quitAddress`. It can be called only by the admin.
```

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: Typo:
```suggestion
     * @param quitAddress The quitAddress address.
```

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: Keeping `immutable`/`constants` variables in the beginning of the file makes it easier to count storage slots.

```suggestion
    /// @notice PSM GEM token contract address
    DSTokenAbstract public immutable gem;
    /// @notice PSM contract address
    PsmAbstract public immutable psm;

    /// @dev This is declared here so the storage layout lines up with RwaInputConduit.
    DSTokenAbstract private __unused_gov;
    /// @notice Dai token contract address
    DSTokenAbstract public dai;
    /// @notice RWA urn contract address
    address public to;

    /// @notice Addresses with admin access on this contract. `wards[usr]`
    mapping(address => uint256) public wards;
    /// @notice Addresses with push access on this contract. `may[usr]`
    mapping(address => uint256) public may;

    /// @notice Exit address
    address public quitAddress;
```

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: I was thinking and maybe we should give this a better name. I've thought of:
- `quitConduit`
- `exitValve`
- `quitTo`
- `exitConduit`
- `emergencyExit`

Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: `quitTo ` sounds good to me

File: src/conduits/RwaInputConduit3.sol
Comment (nikozalo) [inline src/conduits/RwaInputConduit3.sol:182]: Shouldn't we have a file function for `to` as well?

File: src/conduits/RwaInputConduit3.sol
Comment (nikozalo) [inline src/conduits/RwaInputConduit3.sol]: Why is this function `auth`ed, any reason we don't let the `operator` call this?

Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: you are right, operator the one who should call this

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol:182]: hmm...we don't have file function at all for previous version of input conduit

Comment (nikozalo) [reply src/conduits/RwaInputConduit3.sol:182]: Would it makes sense to be able to set `to` in case we update the Urn? Then we don't need to redeploy InputConduit and thus change the target address for the end user.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol]: We don't have operators in input conduits, only admins and mates.
In output conduits, operators are managed by `hope()/nope()`.

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol:182]: it will make sense only in case of changing the URN, but probably we will have new conduit if have a new URN?

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: I did it callable by `mate`

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol:182]: Ok, I am fine with that

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol]: Alright, I was just worried about the nomenclature we use here, because it can become confusing 😅. All good.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [reply src/conduits/RwaInputConduit3.sol:184]: > Would it makes sense to be able to set to in case we update the Urn?

We should probably stick to `pick` here instead of using the `file()` pattern for compatibility sake.

File: src/conduits/RwaInputConduit3.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.sol]: I really think we shouldn't have this check here.
There's no effect of calling this function when it has no balance other than wasting gas.

If we leave it here, it will behave different than the other conduits, which don't revert when calling `push()` with no balance.

**Review by hbarcelos — CHANGES_REQUESTED** [review]

File: src/conduits/RwaInputConduit3.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.sol]: Done

**Comment by nikozalo** [comment]
LGTM, nice work

**Review by nikozalo — APPROVED** [review]

File: src/conduits/RwaInputConduit3.t.sol
Comment (hbarcelos) [inline src/conduits/RwaInputConduit3.t.sol]: Only a suggestion: you can move all these helper contracts to the end of the file, so the actual test contract (which is the main subject) appears first.

**Review by hbarcelos — APPROVED** [review]
LGTM.

GJ :shipit: 

Just one suggestion in the test file to move the helper contracts to the end, but it's up to you.

File: src/conduits/RwaInputConduit3.t.sol
Comment (nduchak) [reply src/conduits/RwaInputConduit3.t.sol]: Done

**Pull request merged**
